a2s/a2s_async.py

```diff
@@ -1,138 +1,226 @@
+    challenge: int = ...,
+    retries: int = ...,
+    ping: Optional[float] = ...,
+) -> List[Player]:
+    ...
+
+
+@overload
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[RulesProtocol],
+    challenge: int = ...,
+    retries: int = ...,
+    ping: Optional[float] = ...,
+) -> Dict[Union[str, bytes], Union[str, bytes]]:
+    ...
+
+
+async def request_async_impl(
+    conn: A2SStreamAsync,
+    encoding: str,
+    a2s_proto: Type[T],
+    challenge: int = 0,
+    retries: int = 0,
+    ping: Optional[float] = None,
+) -> Union[SourceInfo, GoldSrcInfo, Dict[Union[str, bytes], Union[str, bytes]], List[Player]]:
     send_time = time.monotonic()
     resp_data = await conn.request(a2s_proto.serialize_request(challenge))
     recv_time = time.monotonic()
     # Only set ping on first packet received
     if retries == 0:
         ping = recv_time - send_time
 
-    reader = ByteReader(
-        io.BytesIO(resp_data), endian="<", encoding=encoding)
+    reader = ByteReader(io.BytesIO(resp_data), endian="<", encoding=encoding)
 
     response_type = reader.read_uint8()
     if response_type == A2S_CHALLENGE_RESPONSE:
         if retries >= DEFAULT_RETRIES:
-            raise BrokenMessageError(
-                "Server keeps sending challenge responses")
+            raise BrokenMessageError("Server keeps sending challenge responses")
         challenge = reader.read_uint32()
-        return await request_async_impl(
-            conn, encoding, a2s_proto, challenge, retries + 1, ping)
+        return await request_async_impl(conn, encoding, a2s_proto, challenge, retries + 1, ping)
 
     if not a2s_proto.validate_response_type(response_type):
-        raise BrokenMessageError(
-            "Invalid response type: " + hex(response_type))
+        raise BrokenMessageError("Invalid response type: " + hex(response_type))
 
     return a2s_proto.deserialize_response(reader, response_type, ping)
 
 
 class A2SProtocol(asyncio.DatagramProtocol):
+    __slots__ = (
+        "recv_queue",
+        "error_event",
+        "error",
+        "fragment_buf",
+        "transport",
+    )
+
     def __init__(self):
-        self.recv_queue = asyncio.Queue()
-        self.error_event = asyncio.Event()
-        self.error = None
-        self.fragment_buf = []
+        self.recv_queue: asyncio.Queue[bytes] = asyncio.Queue()
+        self.error_event: asyncio.Event = asyncio.Event()
+        self.error: Optional[Exception] = None
+        self.fragment_buf: List[A2SFragment] = []
 
-    def connection_made(self, transport):
-        self.transport = transport
+    def connection_made(self, transport: asyncio.DatagramTransport) -> None:
+        self.transport: asyncio.DatagramTransport = transport
 
-    def datagram_received(self, packet, addr):
-        header = packet[:4]
-        payload = packet[4:]
+    def datagram_received(self, data: bytes, addr: Tuple[str, int]) -> None:
```

Comment:
Why was packet renamed?

Reply:
This again is part of `pyright`'s strict type checking when a type's methods are overwritten "incorrectly" in their terms.
If you subclass and override a method, the existing parameters must have the same name:-

```py
Method "datagram_received" overrides class "DatagramProtocol" in an incompatible manner
  Parameter 2 name mismatch: base parameter is named "data", override parameter is named "packet"
```
![NVIDIA_Share_gb7ygwRmzh](https://user-images.githubusercontent.com/16031716/212230047-8a4d3151-696e-4c7d-b62e-67a418806d54.png)